src/rnnlm/arpa-sampling.h

```diff
@@ -0,0 +1,131 @@
+// arpa_sampling.h
+
+// Copyright     2016  Ke Li
+
+// See ../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ARPA_SAMPLING_H_
+#define ARPA_SAMPLING_H_
+
+#include <algorithm>
+#include <map>
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <sstream>
+#include <vector>
+
+#include <math.h>
+#include <sys/time.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include "fst/fstlib.h"
+#include "util/common-utils.h"
+#include "lm/arpa-file-parser.h"
+
+namespace kaldi {
+
+class ArpaSampling : public ArpaFileParser {
+ public:
+  friend class ArpaSamplingTest;
+
+  // HistType represents a history
+  typedef std::vector<int32> HistType;
+  // WordToProbsMap represents the words and their probabilities given
+  // an arbitrary history
+  typedef std::unordered_map<int32, std::pair<BaseFloat, BaseFloat> > WordToProbsMap; 
```

Comment:
Any typdefs which the public interface doesn't need should be protected or private.